OpenRA.Mods.Common/Traits/ProductionStarport.cs

```diff
@@ -0,0 +1,144 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Deliver multiple units via skylift. Works with BuildProductionQueue")]
+	public class ProductionStarportInfo : ProductionInfo
+	{
+		[NotificationReference("Speech")]
+		[Desc("Speech notification to play when a unit is delivered.")]
+		public readonly string ReadyAudio = "Reinforce";
+
+		[TranslationReference(optional: true)]
+		[Desc("Text notification to display when a unit is delivered.")]
+		public readonly string ReadyTextNotification = null;
+
+		[FieldLoader.Require]
+		[ActorReference(typeof(AircraftInfo))]
+		[Desc("Cargo aircraft used for delivery. Must have the `" + nameof(Aircraft) + "` trait.")]
+		public readonly string ActorType = null;
+
+		[Desc("Direction the aircraft should face to land.")]
+		public readonly WAngle Facing = new(256);
+
+		[Desc("Offset the aircraft used for landing.")]
+		public readonly WVec LandOffset = WVec.Zero;
+
+		public override object Create(ActorInitializer init) { return new ProductionStarport(init, this); }
+	}
+
+	sealed class ProductionStarport : Production
+	{
+		RallyPoint rp;
+		BulkProductionQueue queue;
+		Actor transport;
+
+		public ProductionStarport(ActorInitializer init, ProductionStarportInfo info)
+			: base(init, info) { }
+
+		protected override void Created(Actor self)
+		{
+			base.Created(self);
+
+			rp = self.TraitOrDefault<RallyPoint>();
+		}
+
+		public bool DeliverOrder(Actor producer, List<ActorInfo> orderedActors, string productionType)
+		{
+			if (IsTraitDisabled || IsTraitPaused)
+				return false;
+			var info = (ProductionStarportInfo)Info;
+			var owner = producer.Owner;
+			var map = owner.World.Map;
+			var waitTickbeforeSpawn = 15;
+			var aircraftInfo = producer.World.Map.Rules.Actors[info.ActorType].TraitInfo<AircraftInfo>();
+			queue = owner.World.ActorsWithTrait<BulkProductionQueue>().First(x => x.Actor.Owner == owner
+				&& x.Trait.GetActorsReadyForDelivery().Equals(orderedActors)).Trait;
+			CPos startPos;
+			WAngle spawnFacing;
+
+			startPos = producer.World.Map.ChooseClosestEdgeCell(producer.Location);
+			spawnFacing = producer.World.Map.FacingBetween(startPos, producer.Location, WAngle.Zero);
+
+			var exit = producer.Info.TraitInfos<ExitInfo>().First();
```

Comment:
```suggestion
			var exit = producer.RandomExitOrDefault(producer.World, productionType);
```

and it should be inside FrameEndTask